lib/dry/system/provider/source.rb

```diff
@@ -37,10 +37,20 @@ class << self
           # @see Dry::System::Provider::SourceDSL
           #
           # @api private
-          def for(name:, group: nil, &block)
-            Class.new(self) { |klass|
+          def for(name:, group: nil, superclass: nil, &block)
+            superclass ||= self
+
+            Class.new(superclass) { |klass|
               klass.source_name name
               klass.source_group group
+
+              name_with_group = group ? "#{group}->#{name}" : name
+              klass.instance_eval <<~RUBY, __FILE__, __LINE__ + 1
+                def name
+                  "#{superclass.name}[#{name_with_group}]"
+                end
+              RUBY
```

Comment:
The metaprogramming around this got a bit hairy, I was unable to find an elegant way to handle this with variable superclasses, so hard-coding it this way was the least bad option.